### DISCUSSION TITLE
Why are x-id parameters added to pre-signed URLs?

### DISCUSSION THREAD

**Discussion by chrisharward-qumulo** [aws/aws-sdk-go-v2#2847] [Q&A] [answered]
Hi,

I'm finding that an x-id query parameter is added to pre-signed urls created with this library, like the following:

https://testbucket.s3.us-west-2.amazonaws.com/sample2.txt?
X-Amz-Algorithm=AWS4-HMAC-SHA256&
X-Amz-Credential=REDACTED/20240130/us-west-2/s3/aws4_request&
X-Amz-Date=20240130T190301Z&
X-Amz-Expires=900&
X-Amz-SignedHeaders=host&
**x-id=GetObject&**
X-Amz-Signature=REDACTED

I'm wondering why this is added, I don't see it in the Amazon specs.

**Comment by RanVaknin** [comment, marked as answer]
Hi @chrisharward-qumulo ,

The x-id parameter is an SDK specific thing, it helps the SDK build the request to the correct S3 endpoint. Since multiple S3 operations might share the same endpoint, the SDK's request building functionality has to differentiate between the various S3 operations sent to the same endpoint. This is where x-id comes into place where the actual operation is represented in the url.

Admittedly, I'm not sure what purpose it serves in the context of presigned URLs. My guess is that because the presigning flow includes the request building flow in it, the generated URL also has this x-id query parameter that has no server side implication.

You can use [this example](https://github.com/aws/aws-sdk-go-v2/issues/2702#issuecomment-2215495235) of how you might want to modify the presigner via middleware to see if you can safely remove it. 

Thanks,
Ran~
